Simplify Button examples

ecraig12345 opened the pull request:
#### Pull request checklist

- [x] Include a change request file using `$ yarn change`

#### Description of changes

Eliminate `getClassNames` usage and style interfaces from the Button examples. While it would be good to have some "all-up" examples which show this approach (assuming it continues to be the recommended approach), I think that in component examples, it's just excessive complexity which makes it harder to understand what one actually needs to do to use the component. 

Update button text to be self-descriptive where possible, rather than having separate labels. This simplifies the examples.

Also use Toggles instead of Checkboxes for the disabled/checked controls on the button example pages since they're more appropriate for an on/off state toggle.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/10163)

jdhuntington commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.Toggle.Example.tsx:
```diff
@@ -1,38 +1,22 @@
-export interface IButtonToggleExampleState {
-  muted: boolean;
-}
+export const ButtonToggleExample: React.FunctionComponent<IButtonProps> = props => {
```
👍  for `FunctionComponent`

jdhuntington commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.Toggle.Example.tsx:
```diff
@@ -1,38 +1,22 @@
+      text={muted ? 'Volume muted' : 'Volume unmuted'}
+      iconProps={{ iconName: muted ? 'Volume0' : 'Volume3' }}
+      // tslint:disable-next-line:jsx-no-lambda
+      onClick={() => {
```
Probably best for this to be a `useCallback` to promote best practices.

ecraig12345 replied:
I'm still not convinced that `useCallback` is actually all that necessary/helpful for the average scenario (see also discussion on https://github.com/OfficeDev/office-ui-fabric-react/pull/10143#discussion_r314177944). I think we should all talk about this in person tomorrow.

dzearing replied:
@ecraig12345 you might be right about `useCallback`. https://kentcdodds.com/blog/usememo-and-usecallback

However, here's where I am still on the fence: passing in constantly mutating values.

Imagine that `DefaultButton` is defined using [React.memo](https://reactjs.org/blog/2018/10/23/react-v-16-6.html#reactmemo). Now passing in the constantly regening callbacks will never benefit from the React.memo scenario.

Maybe the right approach is: don't memo anything in a component; it assumes the caller knows what they're doing. Let the caller memo, and when they do, they also must ensure that they don't pass in constantly mutating values. It's not easy to "just know" that though.

Constantly mutating:

```tsx
<DefaultButton iconProps={{ stuff }} /> 
```

ecraig12345 replied:
I'd seen that article too, which is part of what influenced my thinking on the issue.

Not totally sure what you mean by this part (except the "constantly mutating values," I get that):

> Maybe the right approach is: don't memo anything in a component; it assumes the caller knows what they're doing. Let the caller memo, and when they do, they also must ensure that they don't pass in constantly mutating values. It's not easy to "just know" that though.

jdhuntington commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.Primary.Example.tsx:
```diff
@@ -1,43 +1,16 @@
+      disabled={disabled}
+      checked={checked}
+      // tslint:disable-next-line:jsx-no-lambda
+      onClick={() => alert('Clicked')}
```
Probably best for this to use `useCallback` to promote best practices.

ecraig12345 commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.CommandBar.Example.tsx:
```diff
@@ -1,70 +1,34 @@
+        iconProps={{ iconName: 'Add' }}
+        text="New item"
+        // Set split=true to render a SplitButton instead of a regular button with a menu
+        // split={true}
```
Either for when the editor is live or just for someone looking at the example code (which is already possible)

ecraig12345 commented:
At @dzearing's request I switched the examples to be self-descriptive where possible and removed the labels. I also moved all the constantly-mutating objects to consts (based on other discussions). @khmakoto Could you please take another look?

khmakoto commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.CommandBar.Example.tsx:
```diff
@@ -1,34 +1,47 @@
+};
+const addIcon: IIconProps = { iconName: 'Add' };
+const mailIcon: IIconProps = { iconName: 'Mail' };
+const stackStyles: Partial<IStackStyles> = { root: { height: 44 } };
```
Any reason why we have to specify `Partial` here instead of just `IStackStyles`?

ecraig12345 replied:
I got in the habit of doing this since some style interfaces have required members, though it may not actually be needed here.

khmakoto commented on packages/office-ui-fabric-react/src/components/Button/examples/Button.ContextualMenu.Example.tsx:
```diff
@@ -1,116 +1,56 @@
-      </div>
-    );
-  }
+function _getMenu(menuProps: IContextualMenuProps): JSX.Element {
```
Build is complaining because `menuProps` here shadows the `menuProps` variable declared in line 10.
